### PR TITLE
Configurable max fork length, tests, bug-fixing

### PR DESCRIPTION
**Pull request by asolovyov**

### CODE REVIEW

File: scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala [hunk 1]
```diff
@@ -89,26 +92,27 @@ class HistorySynchronizer(application: Application) extends ViewSynchronizer wit
...
 
-      lastUpdate = System.currentTimeMillis()
+      lastUpdate = Some(System.currentTimeMillis())
       val common = blockIds.head
       log.debug(s"Got blockIds: ${blockIds.map(id => Base58.encode(id))}")
 
       val toDownload = blockIds.tail.filter(b => !application.history.contains(b))
       if (application.history.contains(common) && toDownload.nonEmpty) {
-        Try(application.blockStorage.removeAfter(common)) //todo we don't need this call for blockTree
+        application.blockStorage.removeAfter(common)
```
Comment (catena2w) [inline scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala]: Try is required here for BlokTree storage

File: scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala [hunk 2]
```diff
@@ -138,12 +142,13 @@ class HistorySynchronizer(application: Application) extends ViewSynchronizer wit
             if (idx == 0) {
               val toProcess = updBlocks.takeWhile(_._2.isDefined).map(_._2.get)
               log.info(s"Going to process ${toProcess.size} blocks")
-              toProcess.find(bp => !processNewBlock(bp, local = false)).foreach { case failedBlock =>
-                log.warn(s"Can't apply block: ${failedBlock.json}")
-                if (history.lastBlock.uniqueId sameElements failedBlock.referenceField.value) {
-                  connectedPeer.handlerRef ! PeerConnectionHandler.Blacklist
-                }
-                gotoSyncing()
+              toProcess.find(!processNewBlock(_, local = false)).foreach {
```
Comment (catena2w) [inline scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala]: Why have you changed this?

File: scorex-basics/src/main/scala/scorex/network/peer/PeerDatabaseImpl.scala [hunk 3]
```diff
@@ -21,18 +22,30 @@ class PeerDatabaseImpl(settings: Settings, filename: Option[String]) extends Pee
...
+      database.commit()
+    }
   }
 
   override def blacklistPeer(address: InetSocketAddress): Unit = {
-    whitelistPersistence.remove(address)
-    if (!isBlacklisted(address)) blacklist += address.getHostName -> System.currentTimeMillis()
+    if (!isBlacklisted(address)) {
+      blacklist += address.getHostName -> System.currentTimeMillis()
+      new util.ArrayList(whitelistPersistence.keyList())
```
Comment (catena2w) [inline scorex-basics/src/main/scala/scorex/network/peer/PeerDatabaseImpl.scala]: You don't need util.ArrayList here, just

      whitelistPersistence.keyList()
        .filter(_.getHostName == address.getHostName)
        .foreach(whitelistPersistence.remove(_))

Comment (asolovyov) [inline scorex-basics/src/main/scala/scorex/network/peer/PeerDatabaseImpl.scala]: I started from this version, and it did not work - some of removed elements in fact retained in the whitelistPersistence which causes test fails etc. The possible reason is modification of the collection we iterate through. From the Java Iterator.remove() docs: "The behaviour of an iterator is unspecified if the underlying collection is modified while the iteration is in progress in any way other than by calling this method". Since we have here a Java collection, I have decided to use a quick Java-style approach - just to create a copy of the collection.

File: scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala (hunk 2 above)
Comment (asolovyov) [inline scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala]: Just a code formatting, no changes in the functionality. If you decide to rollback the change, I would ask to keep  the placeholder ("_") in lieu of the "bp =>" - that was the main reason not to rollback it.

File: scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala (hunk 1 above)
Comment (asolovyov) [inline scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala]: Yes, I understand this. My concern is this way we "eat" (ignore) all exceptions coming from removeAfter(), even those which are not related to the unsupported operation. I have changed BlockStorage.removeAfter() respectively. Maybe it makes sense in implement this in another way, but ignoring exceptions at this point is a bit risky.

File: scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala [hunk 4]
```diff
@@ -172,10 +185,11 @@ class HistorySynchronizer(application: Application) extends ViewSynchronizer wit
     syncing
   }
 
-  private def gotoGettingExtension(betterScore: BigInt, witnesses: Seq[ConnectedPeer]): Unit = {
+  private def gotoGettingExtension(witnesses: Seq[ConnectedPeer]): Unit = {
     log.debug("Transition to gettingExtension")
     blockGenerator ! StopGeneration
-    context become gettingExtension(betterScore, witnesses)
+    context become gettingExtension(witnesses)
+    gettingExtension(witnesses)
```
Comment (kushti) [inline scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala]: Why  this line?

Comment (asolovyov) [inline scorex-basics/src/main/scala/scorex/network/HistorySynchronizer.scala]: Well, I just tried to find a cause of the problem I have fixed, and I have decided to add this line as for other "goto"s, and since all tests (including test net integration) are passed. 

But, TBH, for me this is the most confusion part of Scorex - why to call the a receive function while processing a message?  We have become() call anyway changing behaviour in this way.
